Objects/listobject.c

```diff
@@ -3536,34 +3546,34 @@ static PyObject *
 listreviter_next(PyObject *self)
 {
     listreviterobject *it = (listreviterobject *)self;
-    PyObject *item;
-    Py_ssize_t index;
-    PyListObject *seq;
-
     assert(it != NULL);
-    seq = it->it_seq;
-    if (seq == NULL) {
-        return NULL;
-    }
+    PyListObject *seq = it->it_seq;
     assert(PyList_Check(seq));
 
-    index = it->it_index;
-    if (index>=0 && index < PyList_GET_SIZE(seq)) {
-        item = PyList_GET_ITEM(seq, index);
-        it->it_index--;
-        return Py_NewRef(item);
+    Py_ssize_t index = LOAD_SSIZE(it->it_index);
+    if (it->it_seq == NULL || index < 0) {
+        return NULL;
     }
-    it->it_index = -1;
+    PyObject *item = list_get_item_ref(seq, index);
+    if (item != NULL) {
+        STORE_SSIZE(it->it_index, index - 1);
+        return item;
+    }
+#ifndef Py_GIL_DISABLED
     it->it_seq = NULL;
     Py_DECREF(seq);
+#endif
+    STORE_SSIZE(it->it_index, -1);
     return NULL;
 }
 
 static PyObject *
 listreviter_len(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
     listreviterobject *it = (listreviterobject *)self;
-    Py_ssize_t len = it->it_index + 1;
+    Py_ssize_t index = LOAD_SSIZE(it->it_index);
+    Py_ssize_t len;
+    STORE_SSIZE(len, index + 1);
```

Comment:
`len` is a local variable so we don't need to use atomic operations.

```suggestion
    Py_ssize_t len = index + 1;
```